### PR TITLE
REPLACE react-datetime with native date/time pickers

### PR DESCRIPTION
**Pull request by ndelangen**
Issue: #4135

## What I did

I replaced the react-datetime components by a custom component using the native date & time picker provided by the browser.

Effects are:
- UX is dependent on the browser
- We no longer have a dependency on moment, so the bundle is smaller

This is based on 3.4 branch

I will do a similar thing for master

### CODE REVIEW

**Comment by ndelangen** [comment]
```
FAIL  lib/core/src/client/preview/client_api.test.js
  ● preview.client_api › storiesOf › add › should replace stories when adding the same story

    expect(received).toBe(expected) // Object.is equality
    
    Expected value to be:
      [Function mockConstructor]
    Received:
      [Function mockConstructor]
    
    Difference:
    
    Compared values have no visual difference.

      312 |           // v3 returns the same function we passed in
      313 |           if (jest.isMockFunction(entry.stories[0].render)) {
    > 314 |             expect(entry.stories[0].render).toBe(stories[0]);
      315 |           } else {
      316 |             expect(entry.stories[0].render()).toBe('story2');
      317 |           }
      
      at Object.<anonymous> (lib/core/src/client/preview/client_api.test.js:314:46)
```

I think this might be an existing problem with 3.4 ?

I didn't change anything related to this 😕

**Comment by ndelangen** [comment]
it is:
https://circleci.com/gh/storybooks/storybook/65125

**Comment by ndelangen** [comment]
Danger also
Netlify's config is missing marko app
And this branch still does cli tests that were removed in master

**Comment by kylemh** [comment]
Wow this will be so much nicer when it's in. Also, one less dependency!

Seeing a lot of issues with trying to make edits to the knob itself via the text input rather than the date picker.

Using up and down arrows to select any input sometimes ends up with the incorrect date being rendered:
arrows-months

Attempting to type out a year results in immediate re-render, the wrong knob value, and the wrong display value:
typing-years

**Comment by ndelangen** [comment]
going to have to make them uncontrolled inputs then. Since right now, when a non-valid date is entered, it's immediately forced into a valid date again, making it impossible to edit freely.

**Comment by ndelangen** [comment]
Hey @kylemh I think I got it working this time.

Might extract this into it's own npm package maybe

**Comment by ndelangen** [comment]
https://deploy-preview-4244--storybooks-official.netlify.com/

**Comment by kylemh** [comment]
Awesome - all of the gif issues resolved. The last issue that I can see is the ability to type and key to non-existent dates, like February 30th of any year.

**Comment by ndelangen** [comment]
That's browser behavior... I dunno maybe they allow it to give you the option of jumping to say 31th of January to 29th of February without having to switch the day first...

The dropdown will only allow you to select real dates in Chrome, but when editing the text, they allow any number < 32 for days in combo with any number < 13 for months. years allow up to 999999, but no negative and chrome will only show the last 4 digits.

**Comment by kylemh** [comment]
Seems fine to me. At the end of the day, this for UI demonstration purposes.

Maybe just ensure we outline the fact that there's no validation beyond what the browser provides in the knobs README for that knob type. Otherwise, looks great to me 👍

**Comment by ndelangen** [comment]
Can I get a review @storybooks/team

**Review by kylemh — APPROVED** [review]

File: addons/knobs/src/components/types/Date/index.js [hunk 1]
```diff
@@ -1,36 +1,119 @@
...
+        <input
+          style={styles}
+          type="time"
+          id={`${name}time`}
+          ref={el => {
+            this.timeInput = el;
+          }}
+          onChange={this.onTimeChange}
+        />
+        {!valid ? <div>invalid</div> : null}
```
Comment (wuweiweiwu) [inline addons/knobs/src/components/types/Date/index.js]: can we just use `!valid &&....`?

**Review by wuweiweiwu — APPROVED** [review]
LGTM! Looks awesome!

File: addons/knobs/package.json [hunk 2]
```diff
@@ -24,7 +24,6 @@
     "moment": "^2.21.0",
```
Comment (libetl) [inline addons/knobs/package.json]: will we be able to get rid of moment ?

Comment (libetl) [reply addons/knobs/package.json]: even in the parent workspace ?

**Comment by pksunkara** [comment]
No other comments other than what others mentioned.

File: addons/knobs/src/components/types/Date/index.js (hunk 1 above)
Comment (Hypnosphi) [reply addons/knobs/src/components/types/Date/index.js]: or even `valid || ...`

Comment (ndelangen) [reply addons/knobs/src/components/types/Date/index.js]: Can you guys find a fixable eslint rule for this?

I personally prefer ternaries over the `&&` or `||` pattern. But I don't care enough about it to fight eslint.

File: addons/knobs/package.json (hunk 2 above)
Comment (libetl) [reply addons/knobs/package.json]: thanks

**Comment by ndelangen** [comment]
Going to close this is favour of the one on master, which should imminently be released as 4.0.0

**Pull request closed without merging**